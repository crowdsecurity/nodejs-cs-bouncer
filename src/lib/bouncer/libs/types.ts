import { LapiClientConfigurations } from 'src/lib/lapi-client/libs/types';
import { RemediationType } from 'src/lib/types';

export type CrowdSecBouncerConfiguration = {
    fallbackRemediation?: RemediationType;
} & LapiClientConfigurations;
