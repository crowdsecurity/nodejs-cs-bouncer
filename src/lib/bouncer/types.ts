import { BOUNCING_LEVEL_DISABLED, BOUNCING_LEVEL_FLEX, BOUNCING_LEVEL_NORMAL } from 'src/lib/bouncer/constants';
import { CacheConfigurations } from 'src/lib/cache/types';
import { LapiClientConfigurations } from 'src/lib/lapi-client/types';
import { WallsOptions } from 'src/lib/rendered/types';
import { Remediation } from 'src/lib/types';

type BouncingLevel = typeof BOUNCING_LEVEL_DISABLED | typeof BOUNCING_LEVEL_FLEX | typeof BOUNCING_LEVEL_NORMAL;

export type CrowdSecBouncerConfigurations = {
    fallbackRemediation?: Remediation; // fallback in case of unknown remediation
    bouncingLevel?: BouncingLevel; // bouncing level
    badIpCacheDuration?: number; // duration in seconds to cache bad IPs
    cleanIpCacheDuration?: number; // duration in seconds to cache clean IPs
    captchaFlowCacheDuration?: number; // duration in seconds to cache captcha flow
    streamMode?: boolean; // stream mode
    wallsOptions?: WallsOptions; // Ban and Captcha walls options
} & LapiClientConfigurations &
    CacheConfigurations;

export type CaptchaSubmission = {
    userPhrase: string;
    ip: string;
    refresh: '0' | '1';
};
