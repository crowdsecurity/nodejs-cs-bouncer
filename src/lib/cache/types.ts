import { CacheAdapter } from 'src/lib/cache/interfaces';
import { DecisionOrigin, CachableDecisionIdentifier, RemediationType, CachableDecisionExpiresAt } from 'src/lib/types';

export type CachableDecisionContent = {
    id: CachableDecisionIdentifier;
    origin: DecisionOrigin;
    expiresAt: CachableDecisionExpiresAt;
    value: RemediationType;
};

export type CachedItem<T = unknown> = {
    key: string;
    content: T;
    ttl?: number; // Time to live in milliseconds
};

export type CachableDecisionItem = CachedItem<CachableDecisionContent[]>;

export type CacheConfigurations = {
    cacheAdapter?: CacheAdapter;
};
