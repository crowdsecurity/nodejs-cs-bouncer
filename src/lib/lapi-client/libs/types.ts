import { DecisionOrigin, DecisionScope } from 'src/lib/types';

export type LapiClientOptions = {
    url: string;
    bouncerApiToken: string;
    userAgent?: string;
    timeout?: number;
};

export type GetDecisionsOptions = {
    isFirstFetch?: boolean;
    origins?: DecisionOrigin[];
    scopes?: DecisionScope[];
    scenariosContaining?: string[];
    scenariosNotContaining?: string[];
};
