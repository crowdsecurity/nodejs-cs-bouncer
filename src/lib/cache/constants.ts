export const CAPTCHA_FLOW = 'captcha_flow';

export const CONFIG = 'config';

export const WARMUP = 'warmed_up';

export const IPV4_BUCKET_KEY = 'range_bucket_ipv4';

export const ORIGINS_COUNT_KEY = 'origins_count';
