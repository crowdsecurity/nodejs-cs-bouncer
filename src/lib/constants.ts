export const SCOPE_IP = 'ip';

export const SCOPE_RANGE = 'range';

export const REMEDIATION_BAN = 'ban';

export const REMEDIATION_CAPTCHA = 'captcha';

export const REMEDIATION_BYPASS = 'bypass';

export const ORIGIN_LISTS = 'lists';

export const ORIGIN_LISTS_SEPARATOR = ':';

export const ORIGIN_CLEAN = 'clean';

export const ID_SEPARATOR = '-';

export const CACHE_SEPARATOR = '_';

export const CACHE_EXPIRATION_FOR_BAD_IP = 120;

export const CACHE_EXPIRATION_FOR_CLEAN_IP = 60;

export const CACHE_EXPIRATION_FOR_CAPTCHA_FLOW = 86400;

export const IP_TYPE_V4 = 'ipv4';

export const IPV4_BUCKET_SIZE = 256;

export const IP_TYPE_V6 = 'ipv6';

export const USER_AGENT_PREFIX = 'nodejs-cs-bouncer';

export enum REFRESH_KEYS {
    NEW = 'new',
    DELETED = 'deleted',
}

export enum BOUNCER_KEYS {
    REMEDIATION = 'remediation',
    ORIGIN = 'origin',
    CAPTCHA_PHRASE = 'captchaPhrase',
}

export const METRICS_TYPE = 'crowdsec-nodejs-bouncer';

export const VERSION = 'v0.1.0';
