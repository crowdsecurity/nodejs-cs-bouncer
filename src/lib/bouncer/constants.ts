import { REMEDIATION_BYPASS, REMEDIATION_CAPTCHA, REMEDIATION_BAN } from 'src/lib/constants';
import { Remediation } from 'src/lib/types';

// From lower priority to higher priority.
export const ORDERED_REMEDIATIONS: Remediation[] = [REMEDIATION_BYPASS, REMEDIATION_CAPTCHA, REMEDIATION_BAN];

export const BOUNCING_LEVEL_DISABLED = 'disabled_bouncing';
export const BOUNCING_LEVEL_FLEX = 'flex_bouncing';
export const BOUNCING_LEVEL_NORMAL = 'normal_bouncing';
