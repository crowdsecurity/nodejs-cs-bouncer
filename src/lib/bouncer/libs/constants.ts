import { REMEDIATION_BYPASS, REMEDIATION_CAPTCHA, REMEDIATION_BAN } from 'src/lib/constants';
import { RemediationType } from 'src/lib/types';

// From lower priority to higher priority.
// @TODO: make this configurable to allow custom remediation types.
export const ORDERED_REMEDIATIONS: RemediationType[] = [REMEDIATION_BYPASS, REMEDIATION_CAPTCHA, REMEDIATION_BAN];
