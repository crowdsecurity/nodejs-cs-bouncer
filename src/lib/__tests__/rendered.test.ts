import { beforeEach, describe, expect, it, jest } from '@jest/globals';

import fs from 'fs';
import { generateTemplate, renderBanWall, renderCaptchaWall } from 'src/lib/rendered';
import { DEFAULT_COLORS, DEFAULT_TEXTS, TEMPLATES_PATH } from 'src/lib/rendered/constants';
import { BaseWallOptions } from 'src/lib/rendered/types';

jest.mock('fs/promises', () => ({
    readFile: jest.fn(),
}));

const readFileSpy = jest.spyOn(fs.promises, 'readFile');

describe('ðŸ“„ Generate Template', () => {
    beforeEach(() => {
        readFileSpy.mockClear();
    });

    it('should generate a base template', async () => {
        // Call the generateTemplate function
        const wall = await generateTemplate('base', {
            colors: DEFAULT_COLORS,
            content: 'Hello World',
            hideCrowdSecMentions: false,
            style: '',
            texts: DEFAULT_TEXTS,
        } as BaseWallOptions);

        expect(wall).toBeTruthy();

        expect(readFileSpy).toHaveBeenCalled();

        // Assert that the readRawFile function was called with the correct arguments
        expect(readFileSpy).toHaveBeenCalledWith(`${TEMPLATES_PATH}/base.ejs`, 'utf8');

        // Assert that the generated template matches the snapshot
        expect(wall).toMatchSnapshot();
    });

    it('should generate a ban template', async () => {
        // Call the generateTemplate function
        const wall = await renderBanWall();

        expect(wall).toBeTruthy();

        expect(readFileSpy).toHaveBeenCalled();

        // Assert that the readRawFile function was called with the correct arguments
        expect(readFileSpy).toHaveBeenNthCalledWith(1, `${TEMPLATES_PATH}/ban.ejs`, 'utf8');
        expect(readFileSpy).toHaveBeenNthCalledWith(2, `${TEMPLATES_PATH}/base.ejs`, 'utf8');

        // Assert that the generated template matches the snapshot
        expect(wall).toMatchSnapshot();
    });

    it('should generate a captcha template', async () => {
        // Call the generateTemplate function
        const wall = await renderCaptchaWall();

        expect(wall).toBeTruthy();

        expect(readFileSpy).toHaveBeenCalled();

        // Assert that the readRawFile function was called with the correct arguments
        expect(readFileSpy).toHaveBeenNthCalledWith(1, `${TEMPLATES_PATH}/captcha.ejs`, 'utf8');
        expect(readFileSpy).toHaveBeenNthCalledWith(2, `${TEMPLATES_PATH}/captcha-css.ejs`, 'utf8');
        expect(readFileSpy).toHaveBeenNthCalledWith(3, `${TEMPLATES_PATH}/base.ejs`, 'utf8');

        // Assert that the generated template matches the snapshot
        expect(wall).toMatchSnapshot();
    });

    it('should be able to customize ban template', async () => {
        const title = 'Custom Title';
        const subtitle = 'Custom Subtitle';

        // Call the generateTemplate function
        const wall = await renderBanWall({
            colors: {
                text: {
                    primary: '#000000',
                    secondary: '#FFFFFF',
                },
                background: {
                    page: '#FFFFFF',
                    container: '#000000',
                },
            },
            texts: {
                title: title,
                subtitle: subtitle,
            },
        });

        expect(wall).toBeTruthy();

        // Check if the custom title and subtitle are present in the generated template
        expect(wall).toContain(title);
        expect(wall).toContain(subtitle);

        // Assert that the generated template matches the snapshot
        expect(wall).toMatchSnapshot();
    });

    it('should be able to customize captcha template', async () => {
        const title = 'Custom Title';
        const subtitle = 'Custom Subtitle';

        const captchaImageTag = '<img src="captcha.png" />';

        // Call the generateTemplate function
        const wall = await renderCaptchaWall({
            colors: {
                text: {
                    primary: '#000000',
                    secondary: '#FFFFFF',
                },
                background: {
                    page: '#FFFFFF',
                    container: '#000000',
                },
            },
            texts: {
                title: title,
                subtitle: subtitle,
            },
            captchaImageTag: captchaImageTag,
        });

        expect(wall).toBeTruthy();

        // Check if the custom title and subtitle are present in the generated template
        expect(wall).toContain(title);
        expect(wall).toContain(subtitle);
        expect(wall).toContain(captchaImageTag);

        // Assert that the generated template matches the snapshot
        expect(wall).toMatchSnapshot();
    });

    it('should display by default CrowdSec mention "Powered by CrowdSec" by default', async () => {
        // Call the generateTemplate function
        const banWall = await renderBanWall();
        const captchaWall = await renderCaptchaWall();

        expect(banWall).toBeTruthy();
        expect(captchaWall).toBeTruthy();

        // Check if the CrowdSec mentions are displayed
        expect(banWall).toContain('Powered by');
        expect(captchaWall).toContain('Powered by');

        // Assert that the generated template matches the snapshot
        expect(banWall).toMatchSnapshot();
        expect(captchaWall).toMatchSnapshot();
    });

    it('should be able to hide CrowdSec mention "Powered by CrowdSec"', async () => {
        // Call the generateTemplate function
        const banWall = await renderBanWall({
            hideCrowdSecMentions: true,
        });
        const captchaWall = await renderCaptchaWall({
            hideCrowdSecMentions: true,
            captchaImageTag: '<img src="captcha.png" />',
        });

        expect(banWall).toBeTruthy();
        expect(captchaWall).toBeTruthy();

        // Check if the CrowdSec mentions are hidden
        expect(banWall).not.toContain('Powered by');
        expect(captchaWall).not.toContain('Powered by');

        // Assert that the generated template matches the snapshot
        expect(banWall).toMatchSnapshot();
        expect(captchaWall).toMatchSnapshot();
    });

    it('should be able to display error message on captcha template', async () => {
        const error = 'Custom Error Message';

        // Call the generateTemplate function
        const wall = await renderCaptchaWall({
            captchaImageTag: '<img src="captcha.png" />',
            texts: {
                error: error,
            },
        });

        expect(wall).toBeTruthy();

        // Check if the error message is displayed
        expect(wall).toContain(error);

        // Assert that the generated template matches the snapshot
        expect(wall).toMatchSnapshot();
    });
});
