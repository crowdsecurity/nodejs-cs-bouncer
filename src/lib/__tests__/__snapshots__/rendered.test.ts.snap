// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ðŸ“„ Generate Template should generate a ban template 1`] = `
"<!doctype html>
<html lang="en">
    <head>
        <title>Oops..</title>
        <meta content="text/html; charset=utf-8" />
        <style>
            * {
                margin: 0;
                padding: 0;
            }
            html {
                height: 100%;
                color: <!- colors.text.primary !>;
            }
            body {
                background: <!- colors.background.page !>;
                font-family: Arial, Helvetica, sans-serif;
                height: 100%;
            }
            .container {
                display: flex;
                align-items: center;
                justify-content: center;
                height: 100%;
            }
            .main {
                background: <!- colors.background.container !>;
                padding: 50px 50px 30px 50px;
                box-shadow: rgba(0, 0, 0, 0.2) 0px 3px 3px -2px, rgba(0, 0, 0, 0.14) 0px 3px 4px 0px, rgba(0, 0, 0, 0.12) 0px 1px 8px 0px;
                border-radius: 10px;
                display: flex;
                align-items: center;
                justify-content: center;
                flex-direction: column;
                z-index: 1;
                position: relative;
            }
            h1 {
                padding: 10px;
            }
            .desc {
                font-size: 1.2em;
                margin-bottom: 30px;
            }
            .powered {
                margin-top: 30px;
                font-size: small;
                color: <!- colors.text.secondary !>;
            }
            .warning {
                width: 40px;
                display: inline-block;
                vertical-align: -4px;
            }
            .logo {
                width: 17px;
                display: inline-block;
                vertical-align: -12px;
                margin-left: 5px;
            }
            a {
                color: <!- colors.text.secondary !>;
            }
            <! print(customCss) !>
        </style>
        
        <script>
            function newImage() {
                document.getElementById('refresh').value = '1';
                document.getElementById('captcha').submit();
            }
        </script>
    </head>

    <body>
        <div class="container">
            <div class="main">
                <h1>ðŸ¤­ Oh!</h1>
<p class="desc">This page is protected against cyber attacks and your IP has been banned by our system.</p>
 
            </div>
        </div>
    </body>
</html>
"
`;

exports[`ðŸ“„ Generate Template should generate a base template 1`] = `
"<!doctype html>
<html lang="en">
    <head>
        <title></title>
        <meta content="text/html; charset=utf-8" />
        <style>
            * {
                margin: 0;
                padding: 0;
            }
            html {
                height: 100%;
                color: <!- colors.text.primary !>;
            }
            body {
                background: <!- colors.background.page !>;
                font-family: Arial, Helvetica, sans-serif;
                height: 100%;
            }
            .container {
                display: flex;
                align-items: center;
                justify-content: center;
                height: 100%;
            }
            .main {
                background: <!- colors.background.container !>;
                padding: 50px 50px 30px 50px;
                box-shadow: rgba(0, 0, 0, 0.2) 0px 3px 3px -2px, rgba(0, 0, 0, 0.14) 0px 3px 4px 0px, rgba(0, 0, 0, 0.12) 0px 1px 8px 0px;
                border-radius: 10px;
                display: flex;
                align-items: center;
                justify-content: center;
                flex-direction: column;
                z-index: 1;
                position: relative;
            }
            h1 {
                padding: 10px;
            }
            .desc {
                font-size: 1.2em;
                margin-bottom: 30px;
            }
            .powered {
                margin-top: 30px;
                font-size: small;
                color: <!- colors.text.secondary !>;
            }
            .warning {
                width: 40px;
                display: inline-block;
                vertical-align: -4px;
            }
            .logo {
                width: 17px;
                display: inline-block;
                vertical-align: -12px;
                margin-left: 5px;
            }
            a {
                color: <!- colors.text.secondary !>;
            }
            <! print(customCss) !>
        </style>
        
        <script>
            function newImage() {
                document.getElementById('refresh').value = '1';
                document.getElementById('captcha').submit();
            }
        </script>
    </head>

    <body>
        <div class="container">
            <div class="main">
                Hello World 
            </div>
        </div>
    </body>
</html>
"
`;

exports[`ðŸ“„ Generate Template should generate a captcha template 1`] = `
"<!doctype html>
<html lang="en">
    <head>
        <title>Oops..</title>
        <meta content="text/html; charset=utf-8" />
        <style>
            * {
                margin: 0;
                padding: 0;
            }
            html {
                height: 100%;
                color: <!- colors.text.primary !>;
            }
            body {
                background: <!- colors.background.page !>;
                font-family: Arial, Helvetica, sans-serif;
                height: 100%;
            }
            .container {
                display: flex;
                align-items: center;
                justify-content: center;
                height: 100%;
            }
            .main {
                background: <!- colors.background.container !>;
                padding: 50px 50px 30px 50px;
                box-shadow: rgba(0, 0, 0, 0.2) 0px 3px 3px -2px, rgba(0, 0, 0, 0.14) 0px 3px 4px 0px, rgba(0, 0, 0, 0.12) 0px 1px 8px 0px;
                border-radius: 10px;
                display: flex;
                align-items: center;
                justify-content: center;
                flex-direction: column;
                z-index: 1;
                position: relative;
            }
            h1 {
                padding: 10px;
            }
            .desc {
                font-size: 1.2em;
                margin-bottom: 30px;
            }
            .powered {
                margin-top: 30px;
                font-size: small;
                color: <!- colors.text.secondary !>;
            }
            .warning {
                width: 40px;
                display: inline-block;
                vertical-align: -4px;
            }
            .logo {
                width: 17px;
                display: inline-block;
                vertical-align: -12px;
                margin-left: 5px;
            }
            a {
                color: <!- colors.text.secondary !>;
            }
            <! print(customCss) !>
        </style>
        <style>
    input {
        margin-top: 10px;
        padding: 10px;
        font-size: 1.1em;
        width: 150px;
    }
    button {
        margin: 30px 0 30px;
        padding: 10px 0;
        font-size: 1.1em;
        background-color: <!- colors.background.button !>;
        color: <!- colors.text.button !>;
        border: none;
        width: 150px;
        border-radius: 5px;
    }
    button:hover {
        background-color: <!- colors.background.button_hover !>;
        cursor: pointer;
    }
    img {
        padding: 10px;
    }
    .error {
        color: <!- colors.text.error_message !>;
        padding: 5px;
    }
</style>

        <script>
            function newImage() {
                document.getElementById('refresh').value = '1';
                document.getElementById('captcha').submit();
            }
        </script>
    </head>

    <body>
        <div class="container">
            <div class="main">
                <h1>
    <svg
        aria-hidden="true"
        focusable="false"
        data-prefix="fas"
        data-icon="exclamation-triangle"
        xmlns="http://www.w3.org/2000/svg"
        viewBox="0 0 576 512"
        class="warning"
    >
        <path
            fill="#f39b2f"
            d="M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
        ></path>
    </svg>
    Hmm, sorry but...
</h1>
<p class="desc">Please complete the security check.</p>


<p>
    <small><button id="refresh_link" onclick="newImage()">refresh image</button></small>
</p>

<form method="POST" id="captcha" action="">
    <input type="text" name="phrase" placeholder="Type here..." autofocus autocomplete="off" />
    <input name="crowdsec_captcha" value="1" />
    <input name="refresh" value="0" id="refresh" />
    

    <button type="submit">CONTINUE</button>
</form>
 
            </div>
        </div>
    </body>
</html>
"
`;
